pkg/statectx/cancel_context.go

```diff
@@ -8,10 +8,11 @@ import (
 	"time"
 )
 
-func newCancelContext() (context.Context, func(err error)) {
+func newCancelContext(ctx context.Context) (context.Context, func(err error)) {
 	result := &cancelContext{
 		done: make(chan struct{}),
 	}
+	result.propagateCancel(ctx)
```

Comment:
Almost copy-paste of context.WithCancel that unfortunately doesn't support custom error codes